platforms/documentation/docs/src/docs/userguide/authoring-builds/plugins/implementing_gradle_plugins_precompiled.adoc

```diff
@@ -0,0 +1,163 @@
+You can provide configuration for your plugin by adding an extension object to this container.
+
+Let's update our  `greetings` example:
+
+.buildSrc/packaging/src/main/kotlin/greetings.gradle.kts
+[source,kotlin]
+----
+// Update to an extension object
+interface GreetingPluginExtension {
+    val message: Property<String>
+}
+
+println("Hello, ${file(project.layout.buildDirectory.file("defaultGreeting.txt")).readText()}")
+
+// Add the 'greeting' extension object to project
+val extension = project.extensions.create<GreetingPluginExtension>("greeting")
+----
+
+You can set the value of the `message` property directly with `extension.message.set("Hi from Gradle,")`.
+
+However, the `GreetingPluginExtension` object becomes available as a project property with the same name as the extension object.
+You can now access `message` like so:
+
+[source,kotlin]
+----
+// Where the<GreetingPluginExtension>() is equivalent to project.extensions.getByType(GreetingPluginExtension::class.java)
+the<GreetingPluginExtension>().message.set("Hi from Gradle")
+----
+
+You can find out more about implementing project extensions in <<custom_gradle_types.adoc#custom_gradle_types,Developing Custom Gradle Types>>.
+
+[[sec:plugin_conventions]]
+== Adding default configuration as conventions
+
+In plugins, you can define default values, also known as *conventions*, using the `project` object.
+
+Convention properties are properties that are initialized with default values but can be overridden:
+
+.buildSrc/packaging/src/main/kotlin/greetings.gradle.kts
+[source,kotlin]
+----
+interface GreetingPluginExtension {
+    val message: Property<String>
+}
+
+val extension = project.extensions.create<GreetingPluginExtension>("greeting")
+
+// Set a default value for 'message'
+extension.message.convention(project.layout.buildDirectory.file("defaultGreeting.txt").get().asFile.readText())
+----
+
+`extension.message.convention(...)` sets a convention for the `message` property of the extension.
+This convention specifies that the value of `message` should default to the content of a file named `defaultGreeting.txt` located in the build directory of the project.
+
+If the `message` property is not explicitly set, its value will be automatically set to the content of `defaultGreeting.txt`.
+
+[[sec:mapping_extension_properties_to_task_properties]]
+== Mapping extension properties to task properties
+
+Using an extension and mapping it to a custom task's input/output properties is common in plugins.
+
+In this example, the message property of the `GreetingPluginExtension` is mapped to the message property of the `GreetingTask` as an input:
+
+.buildSrc/packaging/src/main/kotlin/greetings.gradle.kts
+[source,kotlin]
+----
+interface GreetingPluginExtension {
+    val message: Property<String>
+}
+
+val extension = project.extensions.create<GreetingPluginExtension>("greeting")
+
+// Create a greeting task
+class GreetingTask : DefaultTask() {
```

Comment:
❗ This should be a snippet, so that it remains current.  

It should also create an `abstract class` for the task, so that you don't need to do `projects.objects.property<String>()` explicitly.

The `abstract` part applies to all example tasks and plugins, I didn't remember to check for this until now.